ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -11258,38 +11259,47 @@ private Operator genPostGroupByBodyPlan(Operator curr, String dest, QB qb,
 
     if (qbp.getOrderByForClause(dest) != null) {
       genReduceSink = true;
+      numReducers = 1;
       hasOrderBy = true;
     }
 
+    if (offset > 0 && !hasOrderBy) {
+      HiveConf.StrictChecks.checkOffsetWithoutOrderBy(conf);
+
+      assert limit != null : "OFFSET is always paired with LIMIT";
```

Comment:
Is this Hive specific? Postgres allows
```
select * from test offset 10;
```
Could you please check the standard?

Reply:
I don't have immediate access to the standard and MySQL is the only one I found so far.

Trino 413.

```
trino> SELECT * FROM tpcds.tiny.item OFFSET 10;                                                    
```

Spark 3.4.1.

```
spark-sql (default)> SELECT * FROM item OFFSET 1990;     
```

MySQL 8.

```
mysql> SELECT * FROM item OFFSET 2;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '2' at line 1
```

I guess we can support it by updating the grammar and steps related to OFFSET.

https://github.com/apache/hive/blob/rel/release-4.0.0-alpha-2/parser/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g#L2779-L2785

Reply:
It can be done in a separate patch if necessary.